src/Engine/Data/EnvironmentTexture.cpp

```diff
@@ -0,0 +1,739 @@
+        m_shcoefs[5][col] += hdr[col] * ( c * y * z ) * domega; /* Y_{2-1} = 1.092548 yz */
+        m_shcoefs[7][col] += hdr[col] * ( c * x * z ) * domega; /* Y_{21}  = 1.092548 xz */
+
+        /* L_{20}.  Note that Y_{20} = 0.315392 (3z^2 - 1) */
+        c = 0.315392f;
+        m_shcoefs[6][col] += hdr[col] * ( c * ( 3 * z * z - 1 ) ) * domega;
+
+        /* L_{22}.  Note that Y_{22} = 0.546274 (x^2 - y^2) */
+        c = 0.546274f;
+        m_shcoefs[8][col] += hdr[col] * ( c * ( x * x - y * y ) ) * domega;
+    }
+}
+
+void EnvironmentTexture::tomatrix( void ) {
+    /* Form the quadratic form matrix (see equations 11 and 12 in paper) */
+    int col;
+    float c1, c2, c3, c4, c5;
+    c1 = 0.429043;
+    c2 = 0.511664;
+    c3 = 0.743125;
+    c4 = 0.886227;
+    c5 = 0.247708;
+
+    for ( col = 0; col < 3; col++ ) { /* Equation 12 */
+
+        m_shMatrices[col]( 0, 0 ) = c1 * m_shcoefs[8][col]; /* c1 L_{22}  */
+        m_shMatrices[col]( 0, 1 ) = c1 * m_shcoefs[4][col]; /* c1 L_{2-2} */
+        m_shMatrices[col]( 0, 2 ) = c1 * m_shcoefs[7][col]; /* c1 L_{21}  */
+        m_shMatrices[col]( 0, 3 ) = c2 * m_shcoefs[3][col]; /* c2 L_{11}  */
+
+        m_shMatrices[col]( 1, 0 ) = c1 * m_shcoefs[4][col];  /* c1 L_{2-2} */
+        m_shMatrices[col]( 1, 1 ) = -c1 * m_shcoefs[8][col]; /*-c1 L_{22}  */
+        m_shMatrices[col]( 1, 2 ) = c1 * m_shcoefs[5][col];  /* c1 L_{2-1} */
+        m_shMatrices[col]( 1, 3 ) = c2 * m_shcoefs[1][col];  /* c2 L_{1-1} */
+
+        m_shMatrices[col]( 2, 0 ) = c1 * m_shcoefs[7][col]; /* c1 L_{21}  */
+        m_shMatrices[col]( 2, 1 ) = c1 * m_shcoefs[5][col]; /* c1 L_{2-1} */
+        m_shMatrices[col]( 2, 2 ) = c3 * m_shcoefs[6][col]; /* c3 L_{20}  */
+        m_shMatrices[col]( 2, 3 ) = c2 * m_shcoefs[2][col]; /* c2 L_{10}  */
+
+        m_shMatrices[col]( 3, 0 ) = c2 * m_shcoefs[3][col]; /* c2 L_{11}  */
+        m_shMatrices[col]( 3, 1 ) = c2 * m_shcoefs[1][col]; /* c2 L_{1-1} */
+        m_shMatrices[col]( 3, 2 ) = c2 * m_shcoefs[2][col]; /* c2 L_{10}  */
+        m_shMatrices[col]( 3, 3 ) = c4 * m_shcoefs[0][col] - c5 * m_shcoefs[6][col];
+        /* c4 L_{00} - c5 L_{20} */
+    }
+}
+
+float* EnvironmentTexture::getPixel( float x, float y, float z ) {
+    auto ma  = std::abs( x );
+    int axis = ( x > 0 );
+    auto tc  = axis ? z : -z;
+    auto sc  = y;
+    if ( std::abs( y ) > ma ) {
+        ma   = std::abs( y );
+        axis = 2 + ( y < 0 );
+        tc   = -x;
+        sc   = ( axis == 2 ) ? -z : z;
+    }
+    if ( std::abs( z ) > ma ) {
+        ma   = std::abs( z );
+        axis = 4 + ( z < 0 );
+        tc   = ( axis == 4 ) ? -x : x;
+        sc   = y;
+    }
+    auto s = 0.5f * ( 1.f + sc / ma );
+    auto t = 0.5f * ( 1.f + tc / ma );
+    int is = int( s * m_width );
+    int it = int( t * m_height );
+    return &( m_skyData[axis][4 * ( ( m_height - 1 - is ) * m_width + it )] );
+}
+
+Ra::Engine::Data::Texture* EnvironmentTexture::getSHImage() {
```

Comment:
Replace `float` by `Scalar` for computation, cast to float on the fly prior `char` conversion in pixel assignment.